### PR TITLE
Workaround mobile safari scrolling issues in modals and trays

### LINKED ISSUE #740: iOS Dialogs background still scrolls
# 🐛 Bug Report 

Go to https://react-spectrum.adobe.com/react-spectrum/DialogTrigger.html
Open the first Dialog
Place finger on overlay or underlay and swipe up or down to scroll the page. The background still scrolls.

## 🤔 Expected Behavior

The background should not scroll.

## 💁 Possible Solution

Latest response in bug linked below in context suggests something that may be a little on the extreme side.

> Since iOS Safari prevents attaching the overflow property to the body tag, you can use the position: fixed attribute. Trouble with position: fixed is that when you want to remove the attribute (or make the body scrollable again) the page jumps back up to the top. To fix this, you can use window.scrollTo() to scroll the page back to where the user last scrolled to. Below is the code:
```
    lockBackground = ({menuOpen}) => {
        if (menuOpen) {
            const offsetY = window.pageYOffset;
            document.body.style.top = `${-offsetY}px`;
            document.body.classList.add('js-lock-position');

        } else {
            const offsetY = Math.abs(parseInt(document.body.style.top || 0, 10));
            document.body.classList.remove('js-lock-position');
            document.body.style.removeProperty('top');
            window.scrollTo(0, offsetY || 0);
        }
    }

   .js-lock-position {
     position: fixed;
     overflow: hidden;
   }
```

## 🔦 Context

This is because of https://bugs.webkit.org/show_bug.cgi?id=153852

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          | Safari
| Operating System | iOS

### PR DESCRIPTION
**Pull request by devongovett**
Closes #740

This works around the many issues with scrolling when a modal is open in mobile safari. For example, when the bottom toolbar is not visible, or the keyboard is open, overflow: hidden has no effect. In addition, when tapping or otherwise focusing an input, Safari always attempts to scroll it to be centered, at the expense of other content going off screen.

These are all worked around in various ways. See comment in the code explaining how it works. These hacks are only applied on mobile safari, and the original codepath is taken on all other browsers.

I've also tested this in trays on the combobox mobile branch. I added an `isFixedHeight` prop here, which is basically the same as the `lockHeightToMax` prop added by @LFDanLu on that branch but implemented in a different way.

Added test instructions to the modified Dialog: long content story, so please see that.

Open to ideas on how to write automated tests for this. I'd basically be mocking a ton of DOM APIs that JSDom doesn't have in order to do it, so I'm not totally sure of the value of that. Another reason to run tests in real browsers I guess...

### CODE REVIEW

File: packages/@react-aria/overlays/src/usePreventScroll.ts [hunk 1]
```diff
@@ -26,16 +39,200 @@ export function usePreventScroll(options: PreventScrollOptions = {}) {
...
+    let target = e.target as HTMLElement;
+    if (target.tagName === 'INPUT') {
+      e.preventDefault();
+
+      // Apply a transform to trick Safari into thinking the input is at the top of the page
+      // so it doesn't try to scroll it into view. When tapping on an input, this needs to
+      // be done before the "focus" event, so we have to focus the element ourselves.
+      target.style.transform = 'translateY(-2000px)';
+      target.focus();
+      requestAnimationFrame(() => {
```
Comment (snowystinger) [inline packages/@react-aria/overlays/src/usePreventScroll.ts:134]: do we need to do any of these safely?
is there any chance of needing to cancel the raf? maybe a new dialog opens as the old one closes and it was touch that closed it? I may just not be following closely enough and this isn't possible

File: packages/@react-spectrum/dialog/stories/Dialog.stories.tsx [hunk 2]
```diff
@@ -350,31 +350,50 @@ function renderWithForm({width = 'auto', ...props}) {
...
+        Test instructions on mobile:
+        <ol>
+          <li>Scroll down and open dialog.</li>
+          <li>Tap on the top input. Dialog should resize when the keyboard comes up and remain at the top of the screen. The rest of the page should not scroll (the first time you do this it may scroll slightly due to the bottom toolbar showing).</li>
+          <li>Close the software keyboard.</li>
+          <li>Scroll down and tap on the bottom input. Dialog should resize as before, and focused input should scroll into view.</li>
+          <li>Tap the previous button in the software keyboard. Focus should move to the first input in the dialog. Dialog should not move and page should not scroll.</li>
+          <li>Tap the previous button again. Focus should stay on the first input in the dialog and not move out to the textfield outside the dialog. The page may jump slightly, and come back but there is nothing we can do about this.</li>
+        </ol>
+        Note: all this only works outside the storybook iframe due to VisualViewport limitations.
```
Comment (snowystinger) [inline packages/@react-spectrum/dialog/stories/Dialog.stories.tsx:374]: is this going to be problematic in ExC shell?

Comment (devongovett) [reply packages/@react-spectrum/dialog/stories/Dialog.stories.tsx:374]: yes, but there is nothing we can do about it 😢. we'd need to discuss with them how to mount overlays outside the iframe. luckily I don't think shell has many mobile users...

File: packages/@react-aria/overlays/src/usePreventScroll.ts (hunk 1 above)
Comment (devongovett) [reply packages/@react-aria/overlays/src/usePreventScroll.ts:134]: maaayyybe? In this case I think we'd always want the transform to be reset so it looks right... not sure about the scroll into view ones below

Comment (snowystinger) [reply packages/@react-aria/overlays/src/usePreventScroll.ts:134]: ok, well, it's probably good then until a bug is reported i guess, I've been unable to break it in that manner, i have a feeling it'd involve a bit more going on than our stories have if we were ever to see a break

File: packages/@react-spectrum/dialog/stories/Dialog.stories.tsx (hunk 2 above)
Comment (snowystinger) [reply packages/@react-spectrum/dialog/stories/Dialog.stories.tsx:374]: 😢

File: packages/@react-spectrum/overlays/src/Modal.tsx [hunk 3]
```diff
@@ -90,8 +91,13 @@ let ModalWrapper = forwardRef(function (props: ModalWrapperProps, ref: RefObject
     otherProps.className
   );
 
+  let viewport = useViewportSize();
+  let style: any = {
+    '--spectrum-visual-viewport-height': viewport.height + 'px'
+  };
```
Comment (LFDanLu) [inline packages/@react-spectrum/overlays/src/Modal.tsx:97]: Ahh thats awesome, didn't know we could set up a css variable outside of the css file.

File: packages/@react-spectrum/dialog/stories/Dialog.stories.tsx [hunk 4]
```diff
@@ -350,31 +350,50 @@ function renderWithForm({width = 'auto', ...props}) {
...
-    <div style={{display: 'flex', width, margin: '100px 0'}}>
-      <DialogTrigger defaultOpen>
+    <div style={{display: 'flex', flexDirection: 'column', alignItems: 'start', width, padding: '0 20px'}}>
+      <p>
+        Test instructions on mobile:
+        <ol>
+          <li>Scroll down and open dialog.</li>
+          <li>Tap on the top input. Dialog should resize when the keyboard comes up and remain at the top of the screen. The rest of the page should not scroll (the first time you do this it may scroll slightly due to the bottom toolbar showing).</li>
+          <li>Close the software keyboard.</li>
+          <li>Scroll down and tap on the bottom input. Dialog should resize as before, and focused input should scroll into view.</li>
```
Comment (LFDanLu) [inline packages/@react-spectrum/dialog/stories/Dialog.stories.tsx:370]: I noticed in iOS that if you close the virtual keyboard after this step, the bottom input is scrolled out of sight and the scroll position is moved to somewhere in the middle of the top paragraph. Android retains the scroll position so I assume this has something to do with the workarounds for mobile safari in this pull. Is it possible to keep the original scroll position to keep the bottom input in view? Will try and digest the usePreventScroll code some more to see if I can think up of something.

Comment (LFDanLu) [reply packages/@react-spectrum/dialog/stories/Dialog.stories.tsx:370]: Also perhaps as a consequence of the above, try the following:
1. Open the Dialog
2. Click on the bottom input and close the virtual keyboard. Note the bottom input is no longer in view
3. Scroll to the top input and tap on it. Note that the bottom input is now scrolled into view

File: packages/@react-spectrum/overlays/src/Tray.tsx [hunk 5]
```diff
@@ -68,22 +72,30 @@ let TrayWrapper = forwardRef(function (props: TrayWrapperProps, ref: RefObject<H
...
-  }, []);
+    // When the height is decreasing, and the keyboard is visible
+    // (visual viewport smaller than layout viewport), delay setting
+    // the new max height until after the animation is complete
+    // so that there isn't an empty space under the tray briefly.
+    if (viewport.height < height && viewport.height < window.innerHeight) {
+      timeoutRef.current = setTimeout(() => {
+        setHeight(viewport.height);
+      }, 500);
+    } else {
```
Comment (LFDanLu) [inline packages/@react-spectrum/overlays/src/Tray.tsx:92]: Ah nice fix for the empty space issue under the tray

**Review by LFDanLu — COMMENTED** [review]
Still digesting the changes in usePreventScroll, but left some notes of things that I saw in my testing

File: packages/@react-spectrum/dialog/stories/Dialog.stories.tsx (hunk 4 above)
Comment (devongovett) [reply packages/@react-spectrum/dialog/stories/Dialog.stories.tsx:370]: Ah yeah we could possibly try to restore the scroll position when the keyboard closes as well...

**Review by dannify — APPROVED** [review]

**Pull request merged**